streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java

```diff
@@ -65,17 +66,34 @@
  */
 public final class JoinWindows extends Windows<Window> {
 
+    private static final long DEFAULT_MAINTAIN_DURATION_MS = 24 * 60 * 60 * 1000L; // default: one day
+    private final long maintainDurationMs;
+
     /** Maximum time difference for tuples that are before the join tuple. */
     public final long beforeMs;
     /** Maximum time difference for tuples that are after the join tuple. */
     public final long afterMs;
 
-    private JoinWindows(final long beforeMs, final long afterMs) {
+    private final Duration grace;
+
+    private JoinWindows(final long beforeMs, final long afterMs, final Duration grace, final long maintainDurationMs) {
         if (beforeMs + afterMs < 0) {
             throw new IllegalArgumentException("Window interval (ie, beforeMs+afterMs) must not be negative.");
         }
         this.afterMs = afterMs;
         this.beforeMs = beforeMs;
+        this.grace = grace;
+        this.maintainDurationMs = maintainDurationMs;
+    }
+
+    @SuppressWarnings({"deprecation"}) // removing segments from Windows will fix this
+    private JoinWindows(final long beforeMs,
+                        final long afterMs,
+                        final Duration grace,
+                        final long maintainDurationMs,
+                        final int segments) {
+        this(beforeMs, afterMs, grace, maintainDurationMs);
+        this.segments = segments;
```

Comment:
Unfortunately, `segments` is public, so we can't bypass it.

Reply:
nit: `super.segments` to make clear it's inherited?

Or maybe: add a package private constructor to `Windows` to accept `segment` parameter?